File: entities/stalker.py
# entities/stalker.py
import math
from .base_entity import Entity
from settings import *


class Stalker(Entity):
    def __init__(self, game, x, y):
        super().__init__(game, x, y, (40, 0, 40), "hexagon")
        self.speed = 2.5
        self.flee_mode = False

    def update(self):
        super().update()

        px, py = self.game.player.pos
        p_angle = self.game.player.angle

        dx = self.x - px
        dy = self.y - py
        dist_to_player = math.sqrt(dx * dx + dy * dy)

        angle_to_me = math.atan2(dy, dx) - p_angle
        while angle_to_me > math.pi: angle_to_me -= 2 * math.pi
        while angle_to_me < -math.pi: angle_to_me += 2 * math.pi

        is_seen = -HALF_FOV < angle_to_me < HALF_FOV and dist_to_player < 600

        if is_seen:
            self.color = (60, 0, 60)

            if dist_to_player < 200:
                self.move(math.cos(angle_to_me + p_angle) * 1.0, math.sin(angle_to_me + p_angle) * 1.0)

        else:
            self.color = (30, 0, 30)

            if dist_to_player > 30:
                target_angle = math.atan2(py - self.y, px - self.x)
                self.angle = target_angle

                mx = math.cos(self.angle) * self.speed
                my = math.sin(self.angle) * self.speed
                self.move(mx, my)

File: src/entities/stalker.rs
// src/entities/stalker.rs
use std::f32::consts::PI;
use crate::config::*;
use crate::map::MapHandler;
use super::entity_base::EntityBase;

pub fn update_stalker(
    base: &mut EntityBase,
    player_pos: (f32, f32),
    player_angle: f32,
    map: &mut MapHandler,
    dt: f32,
) {
    let speed = base.speed * dt * 60.0;
    let (px, py) = player_pos;
    let dx = base.x - px;
    let dy = base.y - py;
    let dist = (dx * dx + dy * dy).sqrt();
    let mut angle_to = dy.atan2(dx) - player_angle;
    while angle_to > PI { angle_to -= 2.0 * PI; }
    while angle_to < -PI { angle_to += 2.0 * PI; }
    let is_seen = angle_to.abs() < HALF_FOV && dist < 600.0;

    if is_seen {
        base.color = [120.0/255.0, 50.0/255.0, 120.0/255.0];
        if dist < 200.0 {
            base.move_pos(
                (angle_to + player_angle).cos() * speed,
                (angle_to + player_angle).sin() * speed,
                map,
            );
        }
    } else {
        base.color = [80.0/255.0, 0.0, 80.0/255.0];
        if dist > 30.0 {
            let target = (py - base.y).atan2(px - base.x);
            base.angle = target;
            base.move_pos(
                base.angle.cos() * speed,
                base.angle.sin() * speed,
                map,
            );
        }
    }
}
